### PR TITLE
Graphics Basics and Optimization

### PR DESCRIPTION
**Pull request by danielsaggau**
graphics

### CODE REVIEW

**Comment by mllg** [comment]
Thanks for the PR. Some comments:

* mlr3verse.png is >4MB. A lower resolution would be sufficient.
* If we have SVGs, we should use them. To also allow building of PDFs, also store SVGs as PDF (convert with svg2pdf/inkscape/...). `knitr` can automatically choose the appropriate file type, depending on output format.
* Raster graphics are only favorable if the vectorized image is large, e.g. plots with hundreds of thousands of points.

**Comment by mllg** [comment]
I'm not sure if we have the necessary copyrights for the following images:

* [ ] Iteration.png
* [ ] ml_workflow.png
* [ ] model_fit_tuning.svg
* [ ] nested_resampling.png
* [ ] overview_training.png
* [ ] tuning.png


@berndbischl @mb706 can you have a look? Some images seem to be from a lecture or R course.

**Comment by mb706** [comment]
* nested_resampling.png: has been part of mlr3book for 7 months
* tuning.png, overview_training.png, model_fit_tuning.svg: I believe @juliambr her Hiwi created this?
* Iteration.png: never seen this
* ml_workflow.png: I've seen this in some course before, maybe @ja-thomas knows this

**Comment by mllg** [comment]
@berndbischl @ja-thomas @giuseppec. Can you please have a look? This is blocking me, I want to restructure the book and I want this PR to be merged first.

**Comment by danielsaggau** [comment]
> @berndbischl @ja-thomas @giuseppec. Can you please have a look? This is blocking me, I want to restructure the book and I want this PR to be merged first.

Here just a brief reminder.

**Comment by mllg** [comment]
ping @berndbischl

**Comment by berndbischl** [comment]
1) i am not sure how you expect me to know this very very single pick. as these are taken from my lectures where multiple people work on, wouldnt it have been easier to ask in the mattermost...?

2) i have looked at these, there are all drawn by us:

 tuning.png
 Iteration.png
 model_fit_tuning.svg
 nested_resampling.png
 overview_training.png

this i am not sure, just leave it out, it does not seem useful for mlr3 anyway?

 ml_workflow.png

**Comment by mllg** [comment]
I don't expect you to have the information about every figure, just wanted to have a whitelist. I already reached out to Janek and Giuseppe here. 

@danielsaggau can you update this PR and remove ml_workflow.png?

**Comment by berndbischl** [comment]
> I don't expect you to have the information about every figure, just wanted to have a whitelist. I already reached out to Janek and Giuseppe here

its fne, i did apparently know it for 99%, i guess it was ok, sorry for being slow i should have answered this more quickly :(

**Comment by danielsaggau** [comment]
> I don't expect you to have the information about every figure, just wanted to have a whitelist. I already reached out to Janek and Giuseppe here.
> 
> @danielsaggau can you update this PR and remove ml_workflow.png?

@mllg Yes, I'll finish it asap.

**Comment by danielsaggau** [comment]
@mllg I updated the images, added the pdf and made some minor changes. I added one new image (also from the slides).

ml_performance

**Comment by pat-s** [comment]
I personally find the "resampling" image somewhat chaotic.

Also: Do we need all the .svg imgs to be added to git if we only use .png and .pdf in the book?

**Comment by mllg** [comment]
We could/should use the SVGs for HTML output.

**Comment by mllg** [comment]
`feature_selection.png` is missing in the PR (did you mean `ml_fs.svg`?). The new `make pdf` discovers this.

Regarding graphic file formats:

We are using `knitr::include_graphics()` with `auto_pdf` set to `TRUE`. Thus, we should do the following:

1. Provide the image URI for HTML output as png or svg. Which one to choose depends on multiple criteria:
    - if the svg is very large (basically a complex image or a plot with many data points), a png is preferable.
    - Otherwise use the svg, even if the svg is a few kB larger than the png.
    - Not all svgs are scaled properly. The dimensions can be stored in the svg itself though, see attribute `viewBox`. This is more a nice-to-have. Fall back to png if this is too complicated.

2. For PDF output, the pdf version (same image URI with extension replaced with 'pdf') is automatically used if it exists.
    - If HTML used svg, convert it to pdf.
    - If HTML used png, use png unless we already have a pdf which is **not** created from a raster format such as png.

3. All images should be stored somewhere. I suggest mlr-org/mlr3orga as repository. Only store in one format (svg > pdf > png).


Does that make sense?

**Comment by danielsaggau** [comment]
Ok, so I  editing the pictures. I hope that this is approximately what was in mind with respect to the size. I have two things that might be important to clarify: 

(1) I was not sure what exactly was meant with the mlr-org/mlr3orga. I currently saved them under mlr3book/bookdown/images and created folders for the different types of formats. 

(2) So just to be sure that I understood correctly, I was using the following structure: 
-If circumstances allow, one should use svg 
-If the svg does not work, one should use a pdf ( e.g. create a pdf using the svg)
-If there is no svg at all ( hindering us to use svg or create a pdf) or there is no pdf we need to use a png

**Comment by danielsaggau** [comment]
> `feature_selection.png` is missing in the PR (did you mean `ml_fs.svg`?). The new `make pdf` discovers this.
> 
> Regarding graphic file formats:
> 
> We are using `knitr::include_graphics()` with `auto_pdf` set to `TRUE`. Thus, we should do the following:
> 
> 1. Provide the image URI for HTML output as png or svg. Which one to choose depends on multiple criteria:
>    
>    * if the svg is very large (basically a complex image or a plot with many data points), a png is preferable.
>    * Otherwise use the svg, even if the svg is a few kB larger than the png.
>    * Not all svgs are scaled properly. The dimensions can be stored in the svg itself though, see attribute `viewBox`. This is more a nice-to-have. Fall back to png if this is too complicated.
> 2. For PDF output, the pdf version (same image URI with extension replaced with 'pdf') is automatically used if it exists.
>    
>    * If HTML used svg, convert it to pdf.
>    * If HTML used png, use png unless we already have a pdf which is **not** created from a raster format such as png.
> 3. All source images should be stored somewhere. I suggest mlr-org/mlr3orga as repository. Only store in one format (svg > pdf > png).
> 
> Does that make sense?

To come back to this, I changed the name and format of feature_selection.png to mlr_fs.svg

**Comment by mllg** [comment]
> (1) I was not sure what exactly was meant with the mlr-org/mlr3orga. I currently saved them under mlr3book/bookdown/images and created folders for the different formats.

Ok, fine for now. 

> (2) So just to be sure that I understood correctly, I was using the following structure:
> -If circumstances allow, one should use svg
> -If the svg does not work, one should use a pdf ( e.g. create a pdf using the svg)
> -If there is no svg at all ( hindering us to use svg or create a pdf) or there is no pdf we need to use a png

Exactly.

**Comment by mllg** [comment]
NB: the images must be placed in the same folder, otherwise `knitr::include_graphics()` cannot find the matching PDF for SVG files.

**Comment by danielsaggau** [comment]
> NB: the images must be placed in the same folder, otherwise `knitr::include_graphics()` cannot find the matching PDF for SVG files.

Ok I now placed all images in mlr3book/bookdown/images without separate folders.

File: bookdown/03-optimization.Rmd [hunk 1]
```diff
@@ -1,23 +1,28 @@
...
 In order to tune a machine learning algorithm, one has to specify (1) the [search space](#tuning-optimization), (2) the [optimization algorithm](#tuning-optimization) (aka tuning method) and (3) an evaluation method, i.e., a resampling strategy and a performance measure.
 
-In summary, the sub-chapter on [tuning](#tuning) illustrates how to: 
+A potential tuning process is illustrated in the graphic below: 
+
+```{r 03-optimization-001, echo = FALSE}
+knitr::include_graphics("images/tuning-rescaled.svg")
+```
 
+In summary, the sub-chapter on [tuning](#tuning) illustrates how to:
```
Comment (pat-s) [inline bookdown/03-optimization.Rmd]: empty line afterwards

File: bookdown/03-optimization.Rmd [hunk 2]
```diff
@@ -1,23 +1,28 @@
...
+```
 
+In summary, the sub-chapter on [tuning](#tuning) illustrates how to: 
 * undertake empirically sound [hyperparameter selection](#tuning)
 * select the [optimizing algorithm](#tuning-optimization) 
 * [trigger](#tuning-triggering) the tuning
 * [automate](#autotuner) tuning
 
-This [sub-chapter](#tuning) requires the package `mlr3-tuning`, an extension package which supports hyperparameter tuning.
+The sub-chapter on model tuning requires the package `mlr3-tuning`, an extension package which supports hyperparameter tuning.
```
Comment (pat-s) [inline bookdown/03-optimization.Rmd]: pkg name (still) misspelled, also please use the `r mlr_pkg()` way to link to a pkg.

**Comment by mllg** [comment]
I'm taking over ...

**Pull request merged**